src/linux.rs

```diff
@@ -0,0 +1,373 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::{
+    ffi::CStr,
+    io::{Error, ErrorKind, Read, Result, Write},
+    mem::size_of,
+    net::IpAddr,
+    num::TryFromIntError,
+    ptr, slice,
+};
+
+use libc::{
+    c_int, AF_INET, AF_INET6, AF_NETLINK, AF_UNSPEC, ARPHRD_NONE, IFLA_IFNAME, IFLA_MTU,
+    NETLINK_ROUTE, NLMSG_ERROR, NLM_F_ACK, NLM_F_REQUEST, RTA_DST, RTA_OIF, RTM_GETLINK,
+    RTM_GETROUTE, RTM_NEWLINK, RTM_NEWROUTE, RTN_UNICAST, RT_SCOPE_UNIVERSE, RT_TABLE_MAIN,
+};
+use static_assertions::{const_assert, const_assert_eq};
+
+use crate::{aligned_by, default_err, routesocket::RouteSocket, unlikely_err};
+
+#[allow(
+    clippy::struct_field_names,
+    non_camel_case_types,
+    clippy::too_many_lines
+)]
+mod bindings {
+    include!(env!("BINDINGS"));
+}
+
+use bindings::{ifinfomsg, nlmsghdr, rtattr, rtmsg};
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const AF_INET_U8: u8 = AF_INET as u8;
+const_assert_eq!(AF_INET_U8 as i32, AF_INET);
```

Comment:
You might consider a macro for all of these.

Reply:
Need to grok macros better.